ios/xpc/connection.go

```
@@ -0,0 +1,166 @@
+	if err != nil {
+		return nil, err
+	}
+
+	err = framer.WriteHeaders(http2.HeadersFrameParam{StreamID: rootChannel, EndHeaders: true})
+	if err != nil {
+		return nil, err
+	}
+
+	firstReadFrame, err := framer.ReadFrame()
+	if err != nil {
+		return nil, err
+	} else {
+		if firstReadFrame.Header().Type != http2.FrameSettings {
+			return nil, errors.New("Received unexpected frame from XPC connection")
+		}
+		// TODO : figure out if need to act on this frame
+	}
+
+	err = framer.WriteSettingsAck()
+	if err != nil {
+		return nil, err
+	}
+
+	err = EncodeData(framerDataWriter{
+		Framer:   *framer,
+		StreamID: rootChannel,
+	}, conn.msgId, map[string]interface{}{})
+	if err != nil {
+		return nil, err
+	}
+
+	EncodeEmpty(framerDataWriter{
+		Framer:   *framer,
+		StreamID: rootChannel,
+	}, conn.msgId, 0x201, false) // TODO : figure out why 0x201 (0x1 for always set, 0x200 for ?)
+	if err != nil {
+		return nil, err
+	}
+
+	err = framer.WriteHeaders(http2.HeadersFrameParam{StreamID: replyChannel, EndHeaders: true})
+	if err != nil {
+		return nil, err
+	}
+
+	EncodeEmpty(framerDataWriter{
+		Framer:   *framer,
+		StreamID: replyChannel,
+	}, conn.msgId, 0, true)
+	if err != nil {
+		return nil, err
+	}
+
+	conn.msgId += 1
+
+	_, err = conn.Receive() // TODO : figure out if need to act on this frame
+	if err != nil {
+		return nil, err
+	}
+	_, err = conn.Receive() // TODO : figure out if need to act on this frame
+	if err != nil {
+		return nil, err
+	}
+	_, err = conn.Receive() // TODO : figure out if need to act on this frame
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (c *Connection) Receive() (map[string]interface{}, error) {
+	for true {
+		response, err := c.framer.ReadFrame()
+		if err != nil {
+			return nil, err
+		}
+
+		if response.Header().Type == http2.FrameData {
+			dataFrame := response.(*http2.DataFrame)
+			if dataFrame.Length == 24 { // empty payload
+				return nil, nil
+			}
```

Lets still run the decoder, otherwise you hide a potential error when decoding fails for some reason

---

Done